packages/controller/src/resources/loki/index.ts

```diff
@@ -196,22 +225,56 @@ export function LokiResources(
       ingestion_rate_mb: 500,
       ingestion_burst_size_mb: 2000,
       // enforced in querier
-      max_entries_limit_per_query: 60000 // Per-tenant entry limit per query
+      max_entries_limit_per_query: 60000, // Per-tenant entry limit per query
+      // https://github.com/grafana/loki/blob/main/production/ksonnet/loki/config.libsonnet#L190-L198
+      reject_old_samples: true,
```

Comment:
I see that this is a significant change, and I have also had a peek into the doc you linked. It's a really important design change in Loki that's being discussed there! 

For the record, this is about changing `reject_old_samples` from `false` (default) to `true`:
```
# Whether or not old samples will be rejected.
# CLI flag: -validation.reject-old-samples
[reject_old_samples: <bool> | default = false]
```

Which means that one cannot add log streams to Loki anymore that start arbitrarily long ago in the past.

Two questions:
1) can we maybe make this change in a separate PR / effort? (is this somewhat tied to the introduction to the query frontend?

2) After reading https://grafana.com/docs/loki/latest/design-documents/2021-01-ordering-constraint-removal/#flushing--chunk-creation I am unclear how far the implementation already goes -- is the Loki that we have currently in Opstrace already set up for out-of-order writes (within a single stream)?

My general thinking is that once the Loki in our system _allows for_ out-of-order writes to a single log stream I'd love to reflect this in our test suite very explicitly.

Reply:
> After reading https://grafana.com/docs/loki/latest/design-documents/2021-01-ordering-constraint-removal/#flushing--chunk-creation I am unclear how far the implementation already goes -- is the Loki that we have currently in Opstrace already set up for out-of-order writes (within a single stream)?

> My general thinking is that once the Loki in our system allows for out-of-order writes to a single log stream I'd love to reflect this in our test suite very explicitly.

So, [out of order logs do not seem to be supported upstream](https://github.com/grafana/loki/issues/1544#issuecomment-575772520).  

Then there's this interesting bit in the [blog post](https://grafana.com/docs/loki/latest/design-documents/2021-01-ordering-constraint-removal/#flushing--chunk-creation):

> Loki regularly combines multiple blocks into a chunk and "flushes" it to storage. In order to ensure that reads over flushed chunks remain as performant as possible, we will re-order a possibly-overlapping set of blocks into a set of blocks that maintain monotonically increasing order between them. 

With this in mind, I don't think it makes sense to accept logs with very old timestamps since chunks try to maintain monotonically increasing order for performance reasons.

Another issue I see is accepting data with a timestamp that conflicts with the log retention settings. Genuinely curious about what happens to data that is ingested with timestamps older than the log retention policy. Will it only be available while it's in ingester memory? I assume that as soon as it is flushed to storage and the compactor runs, it marks it for deletion. The test suite we have does not account for how the other components interact. Not saying it should, but it might lead to test things that Loki is not meant to do.

> could we maybe make this change in a separate PR / effort? (is this somewhat tied to the introduction to the query frontend)?

This PR is based on a branch from a few months ago to introduce the query frontend and update the Loki settings to follow what upstream recommends for production deployments. In this case, we should be conservative and follow Loki production settings, and reject samples that might cause an issue with chunk generation. @jgehrcke wdyt?

Reply:
Based on your reply I think I understand that the change to `reject_old_sample` _is_ unrelated to the introduction of the query frontend. 

I think I also understand that you would like to keep it in here.

That's fair.

It's just an important behavioral change that should not be silently piggy-backed onto an 'unrelated' PR, and I think now that we have discussed this here that is already not silent anymore.

> Genuinely curious about what happens to data that is ingested with timestamps older than the log retention policy

I think you bring up a great point about the undefined behavior / inconsistent behavior here. In CI, I think the system is configured to have a retention period of 7 days for logs. Being able to insert log data older than that is simply flawed. I never thought about it that way before, bit it's pretty obvious. To me, this is as of now the best / _strongest_ reason to change the test suite to not used fixed timestamp log entries anymore far from the past.